### PR TITLE
Add more logging info to Kotlin arc startup and StorageProxy messages

### PR DESCRIPTION
**Pull request by mykmartin**

### CODE REVIEW

File: java/arcs/core/storage/StorageProxy.kt [hunk 1]
```diff
@@ -71,7 +71,7 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T> private constru
     val dispatcher: CoroutineDispatcher
         get() = scheduler.asCoroutineDispatcher()
 
-    private val log = TaggedLog { "StorageProxy" }
+    private val log = TaggedLog { "StorageProxy" }.withSuffix { "(key=$storageKey)" }
```
Comment (shaper) [inline java/arcs/core/storage/StorageProxy.kt:74]: This we unfortunately can’t merge as storage key is not safe for non debug logging. Unless we enhance the suffix thing to allow specifying log level?

**Review by shaper — CHANGES_REQUESTED** [review]

File: java/arcs/core/storage/StorageProxy.kt (hunk 1 above)
Comment (cromwellian) [reply java/arcs/core/storage/StorageProxy.kt:74]: You could make a special .debugSuffix() that only appears it to log.debug() messages perhaps.

File: java/arcs/core/util/TaggedLog.kt [hunk 2]
```diff
@@ -11,12 +11,27 @@
...
 /**
- * Allows the user to create a scoped log utility where everything logged will be prepended with the output of
- * [tagBuilder].
+ * Allows the user to create a scoped log utility where everything logged will be prepended with the
+ * output of [tagBuilder]. An optional suffix can be appended using [withSuffix] when constructing.
  */
 class TaggedLog(private val tagBuilder: () -> String) {
+    private var suffixBuilder: (() -> String)? = null
+
     private fun taggedMessageBuilder(messageBuilder: () -> String): () -> String {
```
Comment (cromwellian) [inline java/arcs/core/util/TaggedLog.kt:20]: add a param to this like taggedMessageBuilder(debug: Boolean = false, messageBuilder: () -> String)

File: java/arcs/core/util/TaggedLog.kt [hunk 3]
```diff
@@ -11,12 +11,27 @@
...
-        return { "${tagBuilder()}: ${messageBuilder()}" }
+        return if (suffixBuilder != null) {
```
Comment (cromwellian) [inline java/arcs/core/util/TaggedLog.kt:21]: if (suffixBuilder != null && debug)

File: java/arcs/core/util/TaggedLog.kt (hunk 2 above)
Comment (cromwellian) [reply java/arcs/core/util/TaggedLog.kt:20]: and change TaggedLog.debug() to pass debug = true when calling the taggedMessageBuilder, then withSuffix can contain PII unsafe stuff.

**Review by cromwellian — CHANGES_REQUESTED** [review]
See suggestions for making withSuffix() work only in debug mode

**Comment by mykmartin** [comment]
Will update to make suffixing debug only. That leads to a question, though: is the intent that only `debug` and `verbose` calls be suffixed, or should higher-level log methods have the suffix when the log level itself is debug or verbose?

aka:
```
Log.level = Log.Level.Debug
val log = TaggedLog { ... }.withDebugSuffix { ... }
log.debug { "this should definitely log with the suffix" }
log.warning { "but should this? " }
```

**Comment by cromwellian** [comment]
> Will update to make suffixing debug only. That leads to a question, though: is the intent that only `debug` and `verbose` calls be suffixed, or should higher-level log methods have the suffix when the log level itself is debug or verbose?
> 
> aka:
> 
> ```
> Log.level = Log.Level.Debug
> val log = TaggedLog { ... }.withDebugSuffix { ... }
> log.debug { "this should definitely log with the suffix" }
> log.warning { "but should this? " }
> ```

The higher levels should never be suffixed (unless you want to add withDebugSuffix() vs withSuffix()), because the intent is that information put into withDebugSuffix() should only ever pass into log.debug/log.verbose. You can see a wonky pattern in some places where we do the following:

```
log.info("Foo started")
log.debug("Foo $pii started")
```

If there's a better way, I'm open to it.

**Comment by shaper** [comment]
> > Will update to make suffixing debug only. That leads to a question, though: is the intent that only `debug` and `verbose` calls be suffixed, or should higher-level log methods have the suffix when the log level itself is debug or verbose?
> > aka:
> > ```
> > Log.level = Log.Level.Debug
> > val log = TaggedLog { ... }.withDebugSuffix { ... }
> > log.debug { "this should definitely log with the suffix" }
> > log.warning { "but should this? " }
> > ```
> 
> The higher levels should never be suffixed (unless you want to add withDebugSuffix() vs withSuffix()), because the intent is that information put into withDebugSuffix() should only ever pass into log.debug/log.verbose. You can see a wonky pattern in some places where we do the following:
> 
> ```
> log.info("Foo started")
> log.debug("Foo $pii started")
> ```
> 
> If there's a better way, I'm open to it.

http://b/160251910 tracks cleaning up the wonky pattern.

**Comment by mykmartin** [comment]
```
The higher levels should never be suffixed (unless you want to add withDebugSuffix() vs withSuffix()), because the intent is that information put into withDebugSuffix() should only ever pass into log.debug/log.verbose.
```

My point here is to pin down the desired behaviour when the log level itself is set to debug, for calls above the debug level. So when `Log.level == Log.Level.Debug`, is it ok for `log.error()` to output the potentially sensitive suffix data? Presumably `Log.level` is only ever set to debug in scenarios like now, when we're investigating a bug on internal devices. I don't imagine the debug level would ever be set on a real user's device.

**Comment by shaper** [comment]
> > The higher levels should never be suffixed (unless you want to add withDebugSuffix() vs withSuffix()), because the intent is that information put into withDebugSuffix() should only ever pass into log.debug/log.verbose.
> 
> My point here is to pin down the desired behaviour when the log level itself is set to debug, for calls above the debug level. So when `Log.level == Log.Level.Debug`, is it ok for `log.error()` to output the potentially sensitive suffix data? Presumably `Log.level` is only ever set to debug in scenarios like now, when we're investigating a bug on internal devices. I don't imagine the debug level would ever be set on a real user's device.
> 
> btw I'm not necessarily arguing for this approach, but it may be a potential (if partial) solution to b/160251910

For the binary on user devices, there is a Proguard step that strips out the actual bytecode that produces debug and verbose logging by method name: http://cl/317360086

We could enhance the Proguard-ing. We do need/want to ensure anything debug/verbose log level wise is not present in the production binaries.

**Comment by shaper** [comment]
@ianull to comment further on above, I do like the idea of progress vs. the bug I linked, having both debug and info-or-otherwise sprinkled throughout the code is super unsightly. Thanks for spending time looking at this. I know it can be frustrating walking the PII line, but it would be awesome if we had things generalized and easy to use in a safe way.

For internal purposes logging the debug/verbose bits at error level is in theory fine I think, presuming we Proguard-updated, but could lead to other folks becoming unnecessarily alarmed since it won't be clear it's stripped and it will show in not-debug/verbose log messages.

To be clear, the specific dangerous thing is actually that the suffix includes dynamically constructed values w/ PII like storageKey. withSuffix logging (which I understand is mainly intended as a convenience sort of thing) is ok if it's just logging ints or opaque ids that can't be tied to specific recipes, particles, user info/actions, etc. We used to have, instead of withSuffix, the TaggedLog built dynamically with storage key values embedded, but that was stripped out, you can see some removals here: https://github.com/PolymerLabs/arcs/pull/5552/files#diff-d587bd637b589198509ae237790abd74L58 and more in that PR. That was kind of like a not-generalized "withPrefix", if you will. Personally I'd prefer the suffix approach, having dynamic info embedded in the log prefix made it hard to read and I feel like tail of messages is where dynamic bits should live.

**Comment by mykmartin** [comment]
Thanks for the info Walter. I think adding this sort of info is more than just a convenience. Since we need to rely on logs for on-device debugging, simple checkpoint dumps like "proxy sent onReady" become much less useful once more than one proxy is involved - hence my aim to provide an identifier in the output.

Of course, ensuring PII is not present in prod logs trumps that concern. The Proguard step is interesting; presumably the bytecode associated with generating a string like "key ${storageKey}" would be safe to leave in, assuming the interpolation is not actually executed in prod contexts? Although the safest course of action would not be to rely on that assumption, and just remove the possibility altogether..

I'm a bit confused by this: `For internal purposes logging the debug/verbose bits at error level is in theory fine`. My understanding is that `log.error()` should produce output at all log levels up to `Log.Level.Error`, and my suggestion is that the debug suffix is only ever included in that output when the log level is set to debug or verbose. I'm not proposing that debug info would be present when the log level is at error.

I have an idea for handling this more safely. I'll post that on the tracking bug for discussion.

**Comment by ianull** [comment]
"(key=$storageKey)" should not exist in prod build in any form like const string.
withDebugSuffix sounds fine to constrain the suffix showing only in verbose and debug levels.
For safety, a proguard rule is required to strip out the withDebugSuffix which would also remove the "(key=$storageKey)" due to no effect.

**Comment by mykmartin** [comment]
Thanks Ian.

How specific can the Proguard rule be? I have a WIP CL that adds the following method to `TaggedLog`:

```
    override fun withDebugSuffix(suffixBuilder: () -> String): Logger {
        return if (Log.level > Log.Level.Debug) {
            this
        } else {
            TaggedLogWithDebug(suffixBuilder)
        }
    }
```

At the moment I have `TaggedLogWithDebug` converting all logging calls to use `Log.debug` and appending the sensitive data, on the understanding that this would be stripped by Proguard. However, that would still leave the interpolation strings in the calls to `withDebugSuffix`. Can Proguard be configured to change `withDebugSuffix` to be just `return this`?

**Comment by ianull** [comment]
> Thanks Ian.
> 
> How specific can the Proguard rule be? I have a WIP CL that adds the following method to `TaggedLog`:
> 
> ```
>     override fun withDebugSuffix(suffixBuilder: () -> String): Logger {
>         return if (Log.level > Log.Level.Debug) {
>             this
>         } else {
>             TaggedLogWithDebug(suffixBuilder)
>         }
>     }
> ```
> 
> At the moment I have `TaggedLogWithDebug` converting all logging calls to use `Log.debug` and appending the sensitive data, on the understanding that this would be stripped by Proguard. However, that would still leave the interpolation strings in the calls to `withDebugSuffix`. Can Proguard be configured to change `withDebugSuffix` to be just `return this`?

Afaik Proguard rule can only change return values for `primitive types` thus the `return this` should not work at all.
It'd be difficult/impossible to strip the `withDebugSuffix` out if you're deciding to use a builder pattern.

**Comment by mykmartin** [comment]
> Afaik Proguard rule can only change return values for `primitive types` thus the `return this` should not work at all.
> It'd be difficult/impossible to strip the `withDebugSuffix` out if you're deciding to use a builder pattern.

Ok, that makes it harder. I'll have to think some more about this.

**Comment by mykmartin** [comment]
I'm going to abandon this PR. I have yet to work out a way to have a single log statement (with or without chained methods) that can use a PII-inclusive string which is completely excised by Proguard.

**Pull request closed without merging**